janus_server/src/aggregator.rs

```diff
@@ -850,56 +835,32 @@ impl VdafOps {
                         ReportShareError::VdafPrepError
                     })
             });
-            let prep_trans = step.map(|step| vdaf.prepare_step(step, None));
 
-            report_share_data.push(match prep_trans {
-                Ok(PrepareTransition::Continue(prep_step, prep_msg)) => {
+            report_share_data.push(match init_rslt {
+                Ok((prep_state, prep_share)) => {
                     saw_continue = true;
                     ReportShareData {
                         report_share,
-                        prep_result: PrepareStepResult::Continued(prep_msg.get_encoded()),
-                        agg_state: ReportAggregationState::<A>::Waiting(prep_step, None),
+                        prep_result: PrepareStepResult::Continued(prep_share.get_encoded()),
+                        agg_state: ReportAggregationState::<L, A>::Waiting(prep_state, None),
                     }
                 }
 
-                Ok(PrepareTransition::Finish(output_share)) => {
-                    saw_finish = true;
-                    ReportShareData {
-                        report_share,
-                        prep_result: PrepareStepResult::Finished,
-                        agg_state: ReportAggregationState::<A>::Finished(output_share),
-                    }
-                }
-
-                Ok(PrepareTransition::Fail(err)) => {
-                    warn!(?task_id, nonce = %report_share.nonce, %err, "Couldn't prepare_step report share");
-                    ReportShareData {
-                        report_share,
-                        prep_result: PrepareStepResult::Failed(ReportShareError::VdafPrepError),
-                        agg_state: ReportAggregationState::<A>::Failed(ReportShareError::VdafPrepError),
-                    }
-                },
-
                 Err(err) => ReportShareData {
                     report_share,
                     prep_result: PrepareStepResult::Failed(err),
-                    agg_state: ReportAggregationState::<A>::Failed(err),
+                    agg_state: ReportAggregationState::<L, A>::Failed(err),
                 },
             });
         }
 
         // Store data to datastore.
-        let aggregation_job_state = match (saw_continue, saw_finish) {
-            (false, false) => AggregationJobState::Finished, // everything failed, or there were no reports
-            (true, false) => AggregationJobState::InProgress,
-            (false, true) => AggregationJobState::Finished,
-            (true, true) => {
-                return Err(Error::Internal(
-                    "VDAF took an inconsistent number of rounds to reach Finish state".to_string(),
```

Comment:
This is due to a VDAF API change: this method is for the "aggregate initialization" step. Previously, the VDAF interface required calling `prepare_init` followed by `prepare_step` (with a `None` prepare state); the latter call would return a `PrepareTransition` which could have been `Finish` (though no real VDAFs do so). Now, we only call `prepare_init`, and the return value is a `(PrepareState, PrepareShare)` tuple -- i.e. there is no way for the VDAF interface to signal that initialization led to a finished VDAF run.

That is, this code got removed because the Rust compiler noticed `saw_finish` variable had become read-only (and would always be false); the check was dead code. Note that we still have the equivalent check in the "aggregate continue" code, around L1081.

(Random thought: while it IMO makes sense for VDAFs to not be able to finish without a round of communication, I wonder if DAFs could have been implemented by allowing `prepare_init` to immediately transition to a finished state.)

Reply:
Ah, thanks for the context, SGTM